fix(widgets): reduces BC break possibility in widgets JS code

hypeJunction opened the pull request:
Now calls the init,system event to make the behavior similar to 2.0

Fixes #9523

hypeJunction commented:
@mrclay This issues a deprecation message on all methods of `elgg.ui.widgets`, which is I think is the bigger problem than init.

mrclay commented on js/lib/ui.widgets.js:
```diff
@@ -1,223 +1,35 @@
-elgg.provide('elgg.ui.widgets');
+require(['elgg', 'jquery', 'elgg/widgets'], function (elgg, $, widgets) {
```
This works as long as we inline elgg/widgets and this async function runs immediately. That's probably safe for BC and the cleanup is definitely easier later.

mrclay commented on js/lib/ui.widgets.js:
```diff
@@ -1,223 +1,35 @@
-		});
-	});
-});
+	elgg.register_hook_handler('init', 'system', widgets.init);
```
This should probably go in the widget.

hypeJunction replied:
I want to remove it in 3.0 and make elgg/widgets require elgg/ready

mrclay commented on views/default/elgg.js.php:
```diff
@@ -18,6 +18,8 @@
 // For backwards compatibility...
 echo elgg_view('sprintf.js');
 
+echo elgg_view('elgg/widgets.js');
```
Need a comment before this, like "Inlining this is required for the elgg.ui.widgets API to be defined immediately. Without this, 3rd party plugins calling elgg.ui.widget methods will break."

mrclay commented on views/default/elgg/widgets.js:
```diff
@@ -1,13 +1,208 @@
+		$(this).parent().parent().find('.elgg-body').slideToggle('medium');
+		event.preventDefault();
+	},
+			/**
```
indent

mrclay commented:
This solution makes me nervous, but the cleanliness is a big win and it's probably fine.

mrclay commented:
Commit description maybe should mention that this inlines elgg/widget and relies on "almost synchronous" behavior of `require` since the 3 dependents are already loaded.

hypeJunction committed:
fix(js): deprecate elgg.ui.widgets more reliably

Widgets JS API has been moved to a named elgg/widgets AMD module,
which is inlined in elgg.js to make sure the API is available at runtime.
This allows us to deprecate elgg.ui.widgets in a manner that is less likely
to break plugin code, i.e. ensuring that the API is available when elgg AMD
module is loaded

Fixes #9523

hypeJunction commented:
@mrclay Updated commit message and added inline docs

mrclay commented on views/default/elgg/widgets.js:
```diff
@@ -1,13 +1,223 @@
+ * calls via elgg/widgets module to issue deprecation notices.
+ *
+ * @todo: in 3.x make this an anonymous module
+ * @todo: in 3.x define elgg/ready as a dependency
```
Shouldn't it be elgg/init?

mrclay replied:
Never mind. As long as plugins depend on the module it doesn't matter.

mrclay commented:
LGTM

hypeJunction commented:
Can we get another LGTM and merge?
